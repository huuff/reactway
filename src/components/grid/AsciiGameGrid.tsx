
import { Grid } from "../../grid/grid";
import { Fragment } from "react";

type GameGridProps = {
    grid: Grid
} & { className: string };

const AsciiGameGrid = ({ grid, className }: GameGridProps) => (
    <div className={`${className} font-mono leading-none text-lg`}>
        {[...Array(grid.height)].map((_, y) => (
            <Fragment key={`row-${y}`}>
                {
                    [...Array(grid.width)].map((_, x) => (
                        <span key={`${x}-${y}`}>{grid.get(x, y) ? "X" : "O"}</span>
                    ))
                }
                <br key={`row-break-${y}`} />
            </Fragment>
        ))}
    </div>
)

export default AsciiGameGrid;