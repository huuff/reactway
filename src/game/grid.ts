type Grid = {
    get(x: number, y: number): boolean;
    set(x: number, y: number, state: boolean): void;
}

export type { Grid };