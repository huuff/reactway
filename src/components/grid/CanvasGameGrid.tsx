import { CSSProperties, MouseEventHandler, RefObject, useCallback, useEffect, useMemo, useRef } from "react";
import { Coordinates, GameGridProps, Grid } from "../../grid/grid";
import { useMouseState, useViewportState } from "beautiful-react-hooks";
import { useDarkMode, useDebounce } from "usehooks-ts";
import { Box2D } from "../../util/box-2d";
import tuple from "immutable-tuple";

const CELL_SIZE_MULTIPLIER = 8;

type CanvasGameGridProps = GameGridProps & {
    scrollX?: number;
    scrollY?: number;
};
type Size = {
    width: number;
    height: number;
}

function getBoundingRectOrZeros(ref: RefObject<HTMLElement>) {
    return ref.current?.getBoundingClientRect() ?? { left: 0, top: 0, right: 0, bottom: 0};
}

// TODO: Test it? Can I?
const CanvasGameGrid = ({
    grid, 
    className, 
    toggleCell,
    cellSize,
    scrollX = 0,
    scrollY = 0,
 }: CanvasGameGridProps) => {
    const { isDarkMode } = useDarkMode();
    const cellSizePixels = useMemo(() => CELL_SIZE_MULTIPLIER * cellSize, [cellSize]);
    const gridCanvasRef = useRef<HTMLCanvasElement>(null);

    const { width: windowWidth, height: windowHeight } = useViewportState();
    
    const gridSizePixels = useMemo(() => ({
        width: grid.width * cellSizePixels,
        height: grid.height * cellSizePixels,
    }), [grid.height, grid.width, cellSizePixels]);

    const visibleCellBounds = useVisibleBounds(windowWidth, windowHeight, scrollX, scrollY, gridSizePixels, cellSizePixels)

    useDrawCanvasEffect(gridCanvasRef, grid, cellSizePixels, visibleCellBounds, isDarkMode);

    const hoveredCell = useHoveredCell(grid.width, grid.height, gridCanvasRef, cellSizePixels);
    const onMouseUp = useClickToggleHandler(gridCanvasRef, hoveredCell, grid, toggleCell);

    const highlightedCell = useHighlightedCell(gridCanvasRef, cellSizePixels, hoveredCell);

    const isMouseWithinGrid = useIsMouseWithinGrid(gridCanvasRef);

    return (
        <div onMouseUp={onMouseUp} className={className}>
            { isMouseWithinGrid && (
            <div style={{
                position: "absolute",
                height: `${cellSizePixels}px`,
                width: `${cellSizePixels}px`,
                backgroundColor: "red",
                left: highlightedCell[0],
                top: highlightedCell[1],
                zIndex: 10,
                opacity: 0.5,
            }} />
            )}
            <canvas
                ref={gridCanvasRef}
                className="mx-auto"
                id="grid-canvas"
                height={gridSizePixels.height}
                width={gridSizePixels.width}
            ></canvas>
        </div>
    )
}

function useIsMouseWithinGrid(gridCanvasRef: RefObject<HTMLCanvasElement>): boolean {
    const { clientX, clientY } = useMouseState();
    const boundingRect = gridCanvasRef.current?.getBoundingClientRect();
    if (!boundingRect) {
        return false;
    }

    const { left, top, right, bottom } = boundingRect;

    return clientX >= left && clientX  <= right && clientY >= top && clientY <= bottom;
}

/**
 * Returns a box that indicates the bounds of the visible part of the screen
 */
function useVisibleBounds(
    windowWidth: number,
    windowHeight: number,
    scrollX: number,
    scrollY: number,
    gridSizePixels: Size,
    cellSizePixels: number
): Box2D {
    const visibleBounds = useDebounce(useMemo<Box2D>(() => new Box2D(
        tuple(
            Math.max(scrollX - (windowWidth / 2), 0),
            Math.max(scrollY - (windowHeight / 2), 0)
        ),
        tuple(
            Math.min(scrollX + (windowWidth * 1.5), gridSizePixels.width),
            Math.min(scrollY + (windowHeight * 1.5), gridSizePixels.height),
        ),
    ), [windowHeight, windowWidth, scrollX, scrollY, gridSizePixels]), 250);
    const visibleCellBounds = useMemo<Box2D>(
        () => visibleBounds.divide(cellSizePixels),
        [visibleBounds, cellSizePixels]
    );

    return visibleCellBounds;
}

/**
 * Returns coordinates that match the position of the highlighted cell in the screen,
 * taking into account all displacements and transformations
 */
const useHighlightedCell = (
    gridCanvasRef: RefObject<HTMLCanvasElement>, 
    cellSizePixels: number,
    [hoveredCellX, hoveredCellY]: Coordinates
    ): Coordinates => {

    const { left: leftDisplacement, top } = getBoundingRectOrZeros(gridCanvasRef);

    const topDisplacement = Math.max(top, 0);

    const coordinates = tuple(
        (hoveredCellX * cellSizePixels) + leftDisplacement,
        (hoveredCellY * cellSizePixels) + topDisplacement,
    );

    return coordinates;
}

/**
 * Returns the cell (in the grid data structure) that the cursor is hovering
 */
const useHoveredCell = (
    gridWidth: number,
    gridHeight: number,
    gridCanvasRef: RefObject<HTMLCanvasElement>,
    cellSizePixels: number
): Coordinates => {
    const { clientX, clientY } = useMouseState();
    const { width: viewportWidth, height: viewportHeight } = useViewportState();

    const { left: leftDisplacement, top, } = getBoundingRectOrZeros(gridCanvasRef);

    const gridDisplacementToTheBottom = Math.max(top, 0);

    const mouseX = Math.max(0, clientX - leftDisplacement)
    const mouseY = Math.max(0, clientY - gridDisplacementToTheBottom)

    const result = tuple(
        Math.min(Math.floor((mouseX - (mouseX % cellSizePixels)) / cellSizePixels), gridWidth-1),
        Math.min(Math.floor((mouseY - (mouseY % cellSizePixels)) / cellSizePixels), gridHeight-1),
    );

    return result;
}

function useClickToggleHandler(
    ref: RefObject<HTMLCanvasElement>,
    mouseCell: Coordinates,
    grid: Grid,
    toggleCell: (coordinates: Coordinates) => void,
): MouseEventHandler<HTMLElement> {
    return useCallback((event) => {
        if (grid.contains(mouseCell)) {
            toggleCell(mouseCell);
        }
    }, [ref, mouseCell, grid, toggleCell]);
}

function useDrawCanvasEffect(
    ref: RefObject<HTMLCanvasElement>,
    grid: Grid,
    cellSizePixels: number,
    visibleCellBounds: Box2D,
    isDarkMode: boolean
) {
    useEffect(() => {
        const canvas = ref.current!;

        const ctx = canvas.getContext("2d")!;
        for (const { coordinates: [x, y], isAlive } of grid.boundedIterator(visibleCellBounds)) {
            if (isAlive) {
                ctx.fillStyle = isDarkMode ? "#262626" : "#000000";
            } else {
                ctx.fillStyle = isDarkMode ? "#666666" : "#F0F0F0";
                ctx.strokeRect(x * cellSizePixels, y * cellSizePixels, cellSizePixels, cellSizePixels);
            }
            ctx.fillRect(x * cellSizePixels, y * cellSizePixels, cellSizePixels, cellSizePixels);
        };
    }, [grid, cellSizePixels, visibleCellBounds, isDarkMode])
}

export default CanvasGameGrid;