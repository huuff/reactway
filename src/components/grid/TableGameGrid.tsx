
import { GameGridProps } from "../../grid/grid";
import classNames from "classnames";
import { coordinatesToString } from "../../util/coordinates-to-string";

const TableGameGrid = ({ grid, className, toggleCell }: GameGridProps) => (
    <table className={`${className || ""} w-72 table-fixed`}>
        <tbody>
            {[...Array(grid.height)].map((_, y) => (
                <tr key={`row-${y}`}>
                    {
                        [...Array(grid.width)].map((_, x) => {
                            const coord = coordinatesToString([x, y]);
                            return (
                                <td
                                    key={coord}
                                    data-testid={coord}
                                    onClick={() => toggleCell([x, y])}
                                    className={classNames(
                                        "w-6",
                                        "h-6",
                                        "border",
                                        {
                                            "bg-black": grid.get(x, y),
                                            "bg-white": !grid.get(x, y),
                                            "hover:bg-red-800": grid.get(x, y),
                                            "hover:bg-red-400": !grid.get(x, y),
                                        })}>
                                </td>
                            )
                        })
                    }
                </tr>
            ))}
        </tbody>
    </table>
)

export default TableGameGrid;