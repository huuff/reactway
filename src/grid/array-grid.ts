import { ConwayStrategy, defaultConwayStrategy } from "../game/conway-strategy";
import { Grid, GridCreationSettings, CreateGrid, Coordinates } from "./grid";
import seedrandom from "seedrandom";
import { shouldBeBornAlive } from "../util/birth-function";
import { ReadonlyDeep } from "type-fest";
import tuple from "immutable-tuple";

type InternalGrid = boolean[][];

class ArrayGrid extends Grid {
    private readonly internalGrid: ReadonlyDeep<InternalGrid>;

    public readonly type = "array";

    public readonly height: number;
    public readonly width: number;


    private constructor(internalGrid: InternalGrid, width: number, height: number) {
        super();
        this.internalGrid = internalGrid;
        this.height = height;
        this.width = width;
    }

    static create: CreateGrid = (settings: GridCreationSettings): ArrayGrid => {
        const random = seedrandom(settings.seed);
        return new ArrayGrid([...Array(settings.height + 2)].map((_, y) =>
            [...Array(settings.width + 2)].map((_, x) => {
                if (x === 0 || x === settings.width + 1 || y === 0 || y === settings.height + 1) {
                    return false;
                } else {
                    return shouldBeBornAlive(random, settings.birthFactor);
                }
            })
        ), settings.width, settings.height);
    };


    get = (coordinates: Coordinates): boolean => {
        if (!this.contains(coordinates)) {
            return false;
        } else {
            const [x, y] = coordinates;
            return this.internalGrid[y + 1][x + 1];
        }
    };

    liveNeighbours = ([x, y]: Coordinates): number => {
        if (x === 0 || y === 0 || x >= this.width || y >= this.height) {
            return 0;
        }

        const [actualX, actualY] = [x + 1, y + 1];

        let result = 0;
        if (this.internalGrid[actualY - 1][actualX - 1]) result++;
        if (this.internalGrid[actualY - 1][actualX]) result++;
        if (this.internalGrid[actualY - 1][actualX + 1]) result++;
        if (this.internalGrid[actualY][actualX - 1]) result++;
        if (this.internalGrid[actualY][actualX + 1]) result++;
        if (this.internalGrid[actualY + 1][actualX - 1]) result++;
        if (this.internalGrid[actualY + 1][actualX]) result++;
        if (this.internalGrid[actualY + 1][actualX + 1]) result++;
        return result;
    }

    tick = (strategy: ConwayStrategy): ArrayGrid => {
        let newInternalGrid: boolean[][];
        if (strategy !== defaultConwayStrategy) {
            newInternalGrid = this.internalGrid.map((_, y) =>
                this.internalGrid[y].map((_, x) => {
                    if (x === 0 || y === 0 || x === this.width + 1 || y === this.height + 1) {
                        return false;
                    } else {
                        return strategy(this, tuple(x - 1, y - 1));
                    }
                })
            )
        } else {
            // If the strategy is the default (which is always, unless it's a test)
            // rather than following it, use an optimized version that takes advantage
            // of local, private state to be faster
            newInternalGrid = this.internalGrid.map((_, y) =>
                this.internalGrid[y].map((_, x) => {
                    if (x === 0 || y === 0 || x === this.width + 1 || y === this.height + 1) {
                        return false;
                    } else {
                        const isAlive = this.internalGrid[y][x];

                        let liveNeighbours = 0;
                        if (this.internalGrid[y - 1][x - 1]) liveNeighbours++;
                        if (this.internalGrid[y - 1][x]) liveNeighbours++;
                        if (this.internalGrid[y - 1][x + 1]) liveNeighbours++;
                        if (this.internalGrid[y][x - 1]) liveNeighbours++;
                        if (this.internalGrid[y][x + 1]) liveNeighbours++;
                        if (this.internalGrid[y + 1][x - 1]) liveNeighbours++;
                        if (this.internalGrid[y + 1][x]) liveNeighbours++;
                        if (this.internalGrid[y + 1][x + 1]) liveNeighbours++;

                        if (isAlive) {
                            return liveNeighbours === 2 || liveNeighbours === 3;
                        } else { // Cell is dead
                            return liveNeighbours === 3;
                        }
                    }
                })
            )
        }
        return new ArrayGrid(newInternalGrid, this.width, this.height);
    };

    toggle = ([x, y]: Coordinates): ArrayGrid => {
        const [actualX, actualY] = [x + 1, y + 1];
        const newInternalGrid = this.internalGrid.map((_, y_2) =>
            this.internalGrid[y].map((_, x_2) => !(actualX === x_2 && actualY === y_2)
                ? this.internalGrid[y_2][x_2]
                : !this.internalGrid[y_2][x_2]
            )
        )

        return new ArrayGrid(newInternalGrid, this.width, this.height);
    };
}

export { ArrayGrid };