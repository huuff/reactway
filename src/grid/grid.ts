import range from "lodash/range";
import { ConwayStrategy } from "../game/conway-strategy";
import { iterateCoordinates } from "../util/iterate-coordinates";
import { GridStateWrapper } from "../game/use-grid";
import { Box2D } from "../util/box-2d";
import tuple, {Tuple} from "immutable-tuple";

type GameGridProps = {
    grid: Grid,
    toggleCell: GridStateWrapper["toggleCell"],
    cellSize: number,
    innerRef?: (node: HTMLDivElement | null) => void,
} & { className?: string };

type Coordinates = Tuple;

type GridCreationSettings = {
    readonly height: number;
    readonly width: number;
    readonly birthFactor: number;
    readonly seed: string;
}

type CreateGrid = (settings: GridCreationSettings) => Grid;

type GridType = "array" | "map" | "set";

abstract class Grid {
    abstract readonly type: GridType;
    abstract readonly height: number;
    abstract readonly width: number;
    /**
     * Gives the number of cells that are currently alive
     */
    // TODO: Test it
    abstract readonly population: number;

    /**
     * Returns the state of the cell at the passed-in coordinates:
     * - true: alive
     * - false: dead
     * @param coordinates 
     */
    abstract get(coordinates: Coordinates): boolean;

    /**
     * Returns the next iteration of the game of life according to the given strategy
     * @param strategy 
     */
    abstract tick(strategy: ConwayStrategy): Grid;

    /**
     * Switches the state of the cell at the given coordinates: from alive to dead and
     * from dead to alive
     * @param coordinates 
     */
    abstract toggle(coordinates: Coordinates): Grid;

    contains = ([x, y]: Coordinates): boolean => {
        return x >= 0 && y >= 0 && x <= (this.width - 1) && y <= (this.height - 1);
    };

    liveNeighbours = ([x, y]: Coordinates): number => {
        let result = 0;
        if (this.get(tuple(x-1, y-1))) result++;
        if (this.get(tuple(x-1, y))) result++;
        if (this.get(tuple(x-1, y+1))) result++;
        if (this.get(tuple(x, y-1))) result++;
        if (this.get(tuple(x, y+1))) result ++;
        if (this.get(tuple(x+1, y-1))) result++;
        if (this.get(tuple(x+1, y))) result++;
        if (this.get(tuple(x+1, y+1))) result++;

        return result;
    };

    iterateCoordinates = (f: (coordinates: Coordinates) => void): void =>{
        iterateCoordinates(this.height, this.width, f);
    };

    *[Symbol.iterator](): IterableIterator<Cell> {
        for (const cell of this.boundedIterator(new Box2D(tuple(0,0), tuple(this.width-1, this.height-1)))) {
            yield cell;
        }
    }

    *boundedIterator(bounds: Box2D): IterableIterator<Cell> {
        const [ minX, minY ] = bounds.topLeft;
        const [ maxX, maxY ] = bounds.bottomRight;

        for (const y of range(minY, maxY+1)) {
            for (const x of range(minX, maxX+1)) {
                yield {
                    coordinates: tuple(x, y),
                    isAlive: this.get(tuple(x, y)),
                };
            }
        }
    }

    equals = (other: Grid): boolean  => {
        if (this.height !== other.height || this.width !== other.width) {
            return false;
        }

        for (const {coordinates: [x, y], isAlive} of this) {
                if (isAlive != other.get(tuple(x, y))) {
                    return false;
                }
        }

        return true;
    };
}

type Cell = {
    readonly coordinates: Coordinates;
    readonly isAlive: boolean;
}

export { Grid };
export type { Coordinates, GridCreationSettings, CreateGrid, GameGridProps, GridType };