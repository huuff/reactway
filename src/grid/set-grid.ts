import _ from "lodash";
import seedrandom from "seedrandom";
import wu from "wu";
import { ConwayStrategy } from "../game/conway-strategy";
import { shouldBeBornAlive } from "../util/birth-function";
import { Coordinates, CreateGrid, Grid, GridCreationSettings, GridType } from "./grid";
import { iterateCoordinates } from "../util/iterate-coordinates";
import tuple from "immutable-tuple";

class SetGrid extends Grid {
     // Holds serialized coordinates so as to have decent equality semantics
    private readonly set: Readonly<Set<Coordinates>>;

    public readonly type: GridType = "set";
    public readonly height: number;
    public readonly width: number;

    constructor(tuples: Set<Coordinates>, height?: number, width?: number)
    constructor(tuples: Coordinates[], height?: number, width?: number)
    constructor(
        tuples: Coordinates[] | Set<Coordinates>,
        height?: number,
        width?: number,
    ) {
        super()
        let actualTuples: Coordinates[];
        if (tuples instanceof Set) {
            actualTuples = wu(tuples.values()).toArray();
        } else {
            actualTuples = tuples;
        }
        this.height = height ?? (_(actualTuples).map(([_, y]) => y).max() ?? 0) + 1;
        this.width = width ?? (_(actualTuples).map(([x, _]) => x).max() ?? 0) + 1;
        this.set = new Set(actualTuples);
    }

    static create: CreateGrid = ({
        height,
        width,
        birthFactor,
        seed,
    }: GridCreationSettings): SetGrid => {
        const random = seedrandom(seed);
        const tuples: Coordinates[] = [];

        iterateCoordinates(height, width, ([x, y]) => {
            if (shouldBeBornAlive(random, birthFactor)) {
                tuples.push(tuple(x, y));
            }
        })

        return new SetGrid(tuples, height, width );
    }

    get(x: number, y: number): boolean {
        const result = this.set.has(tuple(x, y));
        return result;
    }

    tick(strategy: ConwayStrategy): SetGrid {
        const nextTuples: Coordinates[] = [];

        this.iterateCoordinates((coordinates) => {
            if (strategy(this, coordinates)) {
                nextTuples.push(coordinates)
            }
        })

        return new SetGrid(nextTuples, this.height, this.width);
    }
    toggle(x: number, y: number): SetGrid {
        const newSet = new Set(this.set);
        if (this.get(x, y)) {
            newSet.delete(tuple(x, y));
        } else {
            newSet.add(tuple(x, y));
        }

        return new SetGrid(
            newSet,
            this.height,
            this.width,
        )
    }
}

export { SetGrid };