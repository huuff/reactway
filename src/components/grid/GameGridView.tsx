import { GameGridProps } from "../../grid/grid";
import { GridViewType } from "../../settings/settings";
import AsciiGameGrid from "./AsciiGameGrid";
import CanvasGameGrid from "./CanvasGameGrid";
import TableGameGrid from "./TableGameGrid";

type GameGridViewProps = GameGridProps & {
    view: GridViewType;
    scrollX: number,
    scrollY: number,
}

const GameGridView = (props: GameGridViewProps) => {
    switch (props.view) {
        case "table":
            return <TableGameGrid {...props} className="mx-auto" />
        case "ascii":
            return <AsciiGameGrid {...props} className="text-center" />
        case "canvas":
            return <CanvasGameGrid {...props} className="mx-auto" />
    }
}

export default GameGridView;