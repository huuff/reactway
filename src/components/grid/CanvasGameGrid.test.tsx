import CanvasGameGrid from "./CanvasGameGrid";
import renderer from "react-test-renderer";
import { Coordinates } from "../../grid/grid";
import { SetGrid } from "../../grid/set-grid";
import tuple from "immutable-tuple";
import { render, screen } from "@testing-library/react";
import "jest-canvas-mock";
import * as beautifulReactHooks from "beautiful-react-hooks";

// Ensure the whole grid is visible
jest.spyOn(beautifulReactHooks, "useViewportState").mockImplementation(() => ({
    height: 10000,
    width: 10000,
    scrollX: 0,
    scrollY: 0,
}));

// https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

const liveCells: Coordinates[] = [tuple(2, 3), tuple(2, 1), tuple(3, 2), tuple(3,3), tuple(4, 2)];
describe("CanvasGameGrid", () => {
    const cellSize = 3;
    const cellSizePixels = cellSize * 8;
    // TODO: Get these from the theme
    const aliveColor = "#000000";
    const deadColor = "#F0F0F0";

    test("live cells are colored as alive", () => {
        // ARRANGE & ACT
        render(<CanvasGameGrid grid={new SetGrid(liveCells)} toggleCell={jest.fn()} cellSize={cellSize}/>);

        // ASSERT
        const canvas: HTMLCanvasElement = screen.getByTestId("canvas");
        const events = canvas.getContext("2d")!.__getEvents();
        
        for (const [x, y] of liveCells) {
            const fillRectEventIndex = events.findIndex(({props}) => props.x === x * cellSizePixels && props.y === y * cellSizePixels);
            const fillRectEvent = events[fillRectEventIndex];
            const fillStyleEvent = events[fillRectEventIndex - 1];
            expect(fillRectEvent.type).toBe("fillRect");
            expect(fillStyleEvent.type).toBe("fillStyle");
            expect(fillStyleEvent.props).toEqual({ value: aliveColor} );

        }
    });

    // TODO: Test dead cells
    // TODO: Test hover
    // TODO: Test toggle

    test("snapshot", () => {
        expect(renderer.create(
            <CanvasGameGrid 
                grid={new SetGrid(liveCells)}
                toggleCell={jest.fn()}
                cellSize={3} 
            />
        ).toJSON()).toMatchSnapshot();
    });
});