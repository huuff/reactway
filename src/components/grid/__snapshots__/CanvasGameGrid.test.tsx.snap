// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasGameGrid snapshot 1`] = `
<div
  style={
    {
      "height": "100vh",
      "position": "relative",
    }
  }
>
  <canvas
    className=""
    height={120}
    id="mouse-canvas"
    style={
      {
        "left": "50%",
        "position": "absolute",
        "transform": "translate(-50%)",
        "zIndex": 10,
      }
    }
    width={150}
  />
  <canvas
    className=""
    height={120}
    id="grid-canvas"
    style={
      {
        "left": "50%",
        "position": "absolute",
        "transform": "translate(-50%)",
      }
    }
    width={150}
  />
</div>
`;
